Fix #2097: implement multi-exploration support for retrieving the top submitted answers for given states

BenHenning opened the pull request:
Implemented a version of stats_services.get_top_state_rule_answers to fetch answers for multiple explorations simultaneously, providing a more efficient means of getting top answers for multiple explorations using a single database call.

wxyxinyu commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
     @classmethod
-    def get_multi(cls, exploration_id, rule_data):
-        """Gets domain objects corresponding to the given rule data.
+    def get_multi(cls, exploration_rule_data_dict):
```
maybe call this `exploration_rule_data_dicts` to emphasize that it is a list? otherwise it just sounds like a dict.

seanlip replied:
Agreed, I was confused by the naming. I like Xinyu's suggestion.

wxyxinyu replied:
hang on. This is actually a dict mapping exploration id to a list of rule data, isn't it?

BenHenning replied:
It was actually a dict mapping to dicts that contained lists. It was confusing. After discussing with Sean, this has been simplified to be a list of tuples instead.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
-    def get_multi(cls, exploration_id, rule_data):
-        """Gets domain objects corresponding to the given rule data.
+    def get_multi(cls, exploration_rule_data_dict):
+        """Gets domain objects given a list of dicts containing an exploration
```
... given a list of dicts, **each** containing an ...

BenHenning replied:
The documentation has been mostly rewritten. PTAL.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
         Args:
-            exploration_id: the exploration id
-            rule_data: a list of dicts, each with the following keys:
+            exploration_rule_data_dict: a list of dicts with an exploration_id
```
I think it would be better to have fixed strings for the keys -- in this case, 'exploration_id' and 'rule_data_list'.

BenHenning replied:
This shouldn't be relevant anymore now that the data structure being passed in is no longer a dict.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
-            rule_data: a list of dicts, each with the following keys:
+            exploration_rule_data_dict: a list of dicts with an exploration_id
+            as the key and a rule_data_list as the value, where each value in
+            the rule_data_list has the following keys:
```
in the rule_data_list **is a dict** with the following keys

BenHenning replied:
Ditto.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
         """
         # TODO(sll): Should each rule_str be unicode instead?
-        answer_log_models = (
+        answer_logs_model_dict = (
```
Normally, get_or_create_multi() returns a list. Can't this be preserved? You're iterating through it anyway...

BenHenning replied:
This behavior has been reverted back to returning a list (``get_or_create_multi`` now returns a list of lists).

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,40 @@ def total_answer_count(self):
+            the rule_data_list has the following keys:
                 (state_name, rule_str).
+
+        Returns a dict with the input exploration IDs as keys mapped to
```
This is confusing me somewhat. AFAICT, a StateRuleAnswerLog object corresponds to a particular rule in a particular state. So, wouldn't you need state_name, rule_str as keys, too?

BenHenning replied:
To avoid clarifying here (since we also clarified on this point offline), PTAL at the new implementation.

wxyxinyu commented:
Is it possible to do `get_top_state_rule_answers_multi` without needing a `top_answer_count_per_rule` argument? Because we actually need the total number of unhandled answers. 

Otherwise, do you think it makes sense to limit that count/limit the number of unhandled answers that will be displayed?

wxyxinyu commented on core/domain/stats_services.py:
```diff
@@ -84,20 +85,48 @@ def get_top_state_rule_answers(
+        exploration_id
+        state_name
+        rule_str_list
+    The number of answers returned is the number of rule spec strings based in
```
suggest: The number of answers returns is at most the number of...

BenHenning replied:
Done.

wxyxinyu commented on core/domain/stats_services_test.py:
```diff
@@ -115,6 +114,138 @@ def test_resolve_answers_for_default_rule(self):
+        self.assertEquals(len(top_answers_dict), 2)
+        self.assertEquals(len(top_answers_dict['eid0']), 1)
+        self.assertEquals(len(top_answers_dict['eid1']), 2)
+
```
Should you also test with multiple states and multiple rules?

BenHenning replied:
I don't think so to the first (since after the latest changes there's no win over testing multiple explorations compared to state names). The second sounds reasonable, though, so added! (I interpreted your second suggestion to mean multiple rule spec strings).

wxyxinyu commented on core/storage/statistics/gae_models.py:
```diff
@@ -84,32 +84,40 @@ class StateRuleAnswerLogModel(base_models.BaseModel):
+            }]})[exploration_id][0]
 
     @classmethod
     def _get_entity_key(cls, unused_exploration_id, entity_id):
```
I'm actually really confused here. So the exploration id isn't used in generating the key?

wxyxinyu replied:
oh nevermind scratch that. I'm really confused by the presence of this function.

seanlip commented:
Per discussion with @BenHenning, suggest refactoring along the following lines: 

    The basic query should be: I give you a bunch of (exploration, state) pairs, and a global filter criterion on the answers. For each state I give you, you give me the answers to it that match the supplied filter criterion (which, for now, is just "answer did not hit the default rule").

BenHenning commented:
Thanks for the review! Restructured per the suggestions and added a new test. PTAL @wxyxinyu and @seanlip.

@wxyxinyu regarding the retrieval number, yes I think this is something that could be changed. I suggest @526avijitgupta does this, though, as it is changing the contract of the current functions and that is outside the scope of this PR (this is extending existing functionality by introducing a new method, rather than changing the underlying functionality).

I don't have a strong opinion on whether the answers should be limited on the dashboard. At the moment, the answers are limited simply because all weren't needed for any existing situations. It might be worthwhile to do so now, though. There doesn't seem to be any extra technical cost in doing so, either. This may be more significant in the future as the next answer storage model does not lend itself to counting quite as well as this one does (and there is more data to deal with).

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,32 @@ def total_answer_count(self):
-            'state_name': state_name,
-            'rule_str': rule_str
-        }])[0]
+        return cls.get_multi([(exploration_id, state_name)], [rule_str])[0][0]
```
Why the second [0]?

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -118,9 +126,17 @@ def get_or_create_multi(cls, exploration_id, rule_data):
-        return entities
+        for (exploration_id, entity) in zip(exploration_ids, entities):
+            entity_dict[exploration_id].append(entity)
+        return entity_dict.values()
```
Are you sure that this preserves the order of the exploration/state pairs that were sent in?

BenHenning replied:
Good points. Updated to guarantee order is consistent and correct for similar exploration IDs and different state names.

seanlip commented on core/domain/stats_services.py:
```diff
@@ -84,20 +82,33 @@ def get_top_state_rule_answers(
+    """
+    answer_logs_list = stats_domain.StateRuleAnswerLog.get_multi(
+        exploration_state_list, rule_str_list)
+    return [
```
(Optional) The indentation here makes things slightly tricky to grok, I think. Maybe this?

```
    return [[
        {
            'value': top_answer[0],
            'count': top_answer[1]
        }
        for answer_log in answer_logs
        for top_answer in answer_log.get_top_answers(
            top_answer_count_per_rule)
    ] for answer_logs in answer_logs_list]
```

BenHenning replied:
Clever, thanks for the suggestion. That's much better.

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,32 @@ def total_answer_count(self):
+                the answers matched to the provided explorations and states
+
+        Returns a list with a list of StateRuleAnswerLog objects for each group
+        of answers matched to each exploration ID-state name.
```
For clarity, can you indicate here that the number of elements in the returned list is the same as the number of elements in exploration_state_list (if that is correct)? I'm starting to get confused when reading stuff like answers, answer_log_models and answer_log_models_list, and I still haven't figured out yet whether the number of elements in the return list is len(exploration_state_list) or len(exploration_state_list) * len(rule_str_list), so it might be worth adding an explicit clarification.

BenHenning replied:
Added a sentence clarifying. The number of elements returned should be ``len(exploration_state_list)``

seanlip commented on core/domain/stats_domain.py:
```diff
@@ -43,28 +43,32 @@ def total_answer_count(self):
-                for answer_log_model in answer_log_models]
+                exploration_state_list, rule_str_list))
+        return [
+            [cls(answer_log_model.answers)
```
Does this inner list have the same number of elements as len(rule_str_list), in the same order? If so, might be worth mentioning that somewhere.

(Though I thought the idea was to just use the rule_strs as filters -- not as individual keys -- and concat everything to get a big log of answers for the given state that match any of the rule filters? Otherwise you're effectively doing a query on the 3-tuple (exp_id, state_name, rule_str).)

BenHenning replied:
The query isn't quite being done on the 3-tuple. The number of ``StateRuleAnswerLog``s returned for each ``(exploration_id, state_name)`` tuple is at most ``len(rule_str_list)``. This was retain previous behavior of returning a list of ``StateRuleAnswerLog``s rather than combining them. I see that returning a combined list might reduce complexity. Plus, you can't easily identify which ``rule_str`` was actually picked. I'm going ahead and doing that.

I don't think any additional clarification is needed in the docstring. Please LMK if you think otherwise.

BenHenning replied:
This actually became a bit more complicated than I anticipated. It turns out that 1 of the 3 callsites of ``get_multi`` actually need to retain this complex behavior. I decided to reintroduce the old ``get_multi`` method and turn this version of it into a new ``get_multi_by_multi_explorations`` to avoid stacking the extra complexity. The answer migration work should lead to these two methods being consolidated (or one of them just being removed altogether).

BenHenning replied:
Edit: you are right about the earlier assessment; it did guarantee a return for each ``rule_str`` in ``rule_str_list``.

seanlip commented on core/domain/stats_services_test.py:
```diff
@@ -49,17 +50,22 @@ class AnalyticsEventHandlersUnitTests(test_utils.GenericTestBase):
+    FUZZY_RULE_TYPE = rule_domain.FUZZY_RULE_TYPE
 
-    def test_record_answer_submitted(self):
+    def _record_answer(
```
Good call on factoring this out!

seanlip commented on core/domain/stats_services_test.py:
```diff
@@ -115,6 +118,150 @@ def test_resolve_answers_for_default_rule(self):
+            'eid', 'sname',
+            [self.FUZZY_RULE_TYPE, self.DEFAULT_RULESPEC_STR], 10)
+        self.assertEquals(len(top_answers), 4)
+        # Fuzzy rules are listed first due to them being listed first in the
```
/cc @wxyxinyu -- we still have fuzzy rules? Might affect #1318 depending on who merges first.

BenHenning replied:
Done.

seanlip commented on core/domain/stats_services.py:
```diff
@@ -84,20 +82,33 @@ def get_top_state_rule_answers(
+    """Returns a list of top answers (by submission frequency) submitted to the
+    given explorations and states which were mapped to any of the rules listed
+    in 'rule_str_list'. The number of answers returned is at most the number of
+    rule spec strings multiplied by top_answer_count_per_rule.
```
Maybe clarify length and order of the return value in relation to the input parameters (similar to above).

BenHenning replied:
Done.

seanlip commented on core/domain/stats_services_test.py:
```diff
@@ -115,6 +118,150 @@ def test_resolve_answers_for_default_rule(self):
+
+    def test_get_top_state_rule_answers_from_multiple_explorations(self):
+        # There are no initial top answers for these explorations.
+        top_answers_dict = stats_services.get_top_state_rule_answers_multi(
```
top_answers_list? Or maybe top_states_answers?

seanlip replied:
Ditto elsewhere (not a dict).

BenHenning replied:
Good call, done.

seanlip commented:
Thanks a lot, @BenHenning! I think this is a lot clearer, though I still have a few questions of clarification.

wxyxinyu commented on core/storage/statistics/gae_models.py:
```diff
@@ -84,32 +84,40 @@ class StateRuleAnswerLogModel(base_models.BaseModel):
-        """Gets or creates entities for the given rules.
+    def get_or_create_multi(cls, exploration_state_list, rule_str_list):
+        """Gets entities given a list of exploration ID and state name tuples,
+        and a list of rule spec strings to filter answers matched to the given
```
maybe "to filter answers matched to each of the given explorations and states"? I was a bit confused about the relation between the explorations/states and the rule_str_list.

BenHenning replied:
Sgtm, changed.

wxyxinyu commented on core/storage/statistics/gae_models.py:
```diff
@@ -84,32 +84,40 @@ class StateRuleAnswerLogModel(base_models.BaseModel):
+            entity_ids += (
+                ['.'.join([
+                    exploration_id, state_name, _OLD_SUBMIT_HANDLER_NAME,
+                    rule_str])[:490]
```
What's the 490 for?

BenHenning replied:
Legacy: we limit the size of answer identifier strings in the old data model.

BenHenning commented:
@wxyxinyu  @seanlip  Okay, these next commits are major changes. I needed to reintroduce the old ``get_multi`` code path, which I really don't like. This is because trying to coerce the new function to return the right order for legacy stats and improvement methods in stats_services is challenging and arguably impossible. This seems like a fine compromise to me because most of this is temporary code, anyway. Added necessary TODOs to remove once possible.

Furthermore, I did end up introducing a test for multiple states as @wxyxinyu suggested. Sean's note about the ordering did reveal to me that different states for the same exploration will behave differently than just exploration state name pairs.

Finally, I decided to go ahead and remove the length argument. So @wxyxinyu, @526avijitgupta will not need to do any work on this before using it.

As an aside: I'm uncomfortable we renamed FuzzyMatches. That probably wasn't a good idea; it's going to significantly increase our answer storage in the old model. That string was NOT trivially changeable and shouldn't have been changed. I think it was changed in a follow-up after I reviewed part of the rules.json changes or I didn't notice it.

Thanks!

BenHenning commented:
FYI: I haven't manually tested answer stats yet. There is one dependent function that does not have automated tests, but it's now using the old code path so it should be a no-op.

seanlip commented:
Can't do a review tonight, sorry --  but I did pick up on the note about
FuzzyMatches. Just to clarify, why exactly is this a problem, especially
since we don't have any of the fuzzy matching stuff running in production
yet? Is it because the name is too long? If so we can try shortening it
(possibly drastically). But if we want to do this, we'd better do it soon,
before the next release...

/cc @wxyxinyu



On Wednesday, June 22, 2016, Ben Henning <notifications@github.com> wrote:

> FYI: I haven't manually tested answer stats yet. There is one dependent
> function that does not have automated tests, but it's now using the old
> code path so it should be a no-op.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2123#issuecomment-227919971>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AKFeyv2F0115iE4M5Z1ingo75mU-Rqukks5qOdhPgaJpZM4I4E7K>
> .
>

BenHenning commented:
@seanlip it's because when we fetch for answers, we create slots for them. Every exploration that's been played since fuzzy rules were introduced have had fuzzy matches entries, along with default (even if no answers were submitted). This wasn't worth addressing since we're moving to a new system. Nonetheless, renaming fuzzy matches is going to result in a lot of new entities being created.

wxyxinyu commented:
@BenHenning do you think we should change it back, then? (And maybe change when answers are migrated)

wxyxinyu commented:
btw, I took a look through the code, I think it looks good.

seanlip commented on core/domain/stats_services.py:
```diff
@@ -77,27 +77,35 @@ def get_state_rules_stats(exploration_id, state_name):
+    in 'rule_str_list' for each exploration ID and state name tuple in
+    exploration_state_list.
+
+    For each exploration ID and state, all answers submitted that match the rule
```
doc nit: that match any of the ...

BenHenning replied:
Done.

seanlip commented:
Code lgtm, too. We should figure out what to do re FuzzyMatches but perhaps that's a separate discussion?

BenHenning commented:
Thanks @wxyxinyu and @seanlip.

@wxyxinyu what do you mean by changing when answers are migrated?

@seanlip I'll add this to the TL meeting notes for Monday. We should discuss this before the next release goes out.
